File: lib/fsutil.ts
// Filesystem + id helpers shared across subcommands. sanitizeName/idOf are moved
// verbatim from dump_articles.ts (behavior unchanged — same per-article filenames).

export function sanitizeName(s: string): string {
  return s.replace(/[^a-zA-Z0-9_-]/g, "_").replace(/_+/g, "_").replace(/^_|_$/g, "");
}

// Stable, human-friendly id for a per-article filename.
export function idOf(r: Record<string, unknown>): string {
  const raw = (r.raw as Record<string, unknown>) ?? {};
  const candidate = (raw.f5_kb_id as string) ||
    (raw.permanentid as string) ||
    (r.uniqueId as string) ||
    (r.title as string) ||
    "article";
  return sanitizeName(candidate).slice(0, 120);
}

export async function readJson<T = unknown>(path: string): Promise<T> {
  return JSON.parse(await Deno.readTextFile(path)) as T;
}

// Pretty-printed JSON with a trailing newline (matches existing output files).
export async function writeJson(path: string, data: unknown): Promise<void> {
  await Deno.writeTextFile(path, JSON.stringify(data, null, 2) + "\n");
}

// Yield each article JSON file in a type dir, skipping `_catalogue.json`/`_index.json`.
export async function* walkArticleFiles(typeDir: string): AsyncGenerator<string> {
  for await (const entry of Deno.readDir(typeDir)) {
    if (entry.isFile && entry.name.endsWith(".json") && !entry.name.startsWith("_")) {
      yield `${typeDir}/${entry.name}`;
    }
  }
}

export async function listTypeDirs(dumpDir: string): Promise<string[]> {
  const dirs: string[] = [];
  for await (const e of Deno.readDir(dumpDir)) {
    // Skip bookkeeping dirs (_pending/, _replaced/, _deleted/): a real type dir is a
    // sanitized type key, which never starts with "_" (sanitizeName strips it).
    if (e.isDirectory && !e.name.startsWith("_")) dirs.push(e.name);
  }
  dirs.sort();
  return dirs;
}

export async function exists(path: string): Promise<boolean> {
  try {
    await Deno.stat(path);
    return true;
  } catch {
    return false;
  }
}

File: f5kb/lib/fsutil.py
"""Filesystem + ID helpers shared across subcommands."""

from __future__ import annotations

import json
import re
import time
from pathlib import Path
from typing import Any, Generator


def sanitize_name(s: str) -> str:
    s = re.sub(r"[^a-zA-Z0-9_-]", "_", s)
    s = re.sub(r"_+", "_", s)
    return s.strip("_")


def id_of(r: dict) -> str:
    """Stable, human-friendly ID for a per-article filename."""
    raw = r.get("raw", {}) or {}
    candidate = (
        raw.get("f5_kb_id")
        or raw.get("permanentid")
        or r.get("uniqueId")
        or r.get("title")
        or "article"
    )
    return sanitize_name(str(candidate))[:120]


def read_json(path: str | Path) -> Any:
    return json.loads(Path(path).read_text(encoding="utf-8"))


def write_json(path: str | Path, data: object) -> None:
    """Pretty-printed JSON with trailing newline (matches existing output files)."""
    p = Path(path)
    p.parent.mkdir(parents=True, exist_ok=True)
    p.write_text(json.dumps(data, indent=2) + "\n", encoding="utf-8")


def walk_article_files(type_dir: str | Path) -> Generator[str, None, None]:
    """Yield each article JSON file in a type dir, skipping _catalogue.json/_index.json."""
    for entry in sorted(Path(type_dir).iterdir()):
        if entry.is_file() and entry.suffix == ".json" and not entry.name.startswith("_"):
            yield str(entry)


def list_type_dirs(dump_dir: str | Path) -> list[str]:
    """Return subdirs not starting with '_' (skips _pending/, _replaced/, _deleted/)."""
    dirs = []
    for entry in Path(dump_dir).iterdir():
        if entry.is_dir() and not entry.name.startswith("_"):
            dirs.append(entry.name)
    return sorted(dirs)


def path_exists(path: str | Path) -> bool:
    return Path(path).exists()


def iso_now() -> str:
    """Current UTC time as ISO 8601 string: 2024-01-15T12:34:56Z"""
    return time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
